### PR TITLE
[Calyx] Add Cell trait.

### PR DESCRIPTION
**Pull request by cgyurgyik**
The `Cell` trait is used to annotate each sub-component within a component. This consists of primitives (e.g. `RegOp`) and component instances. This will be useful for future passes that want to work on `Cell`s, such as resource-sharing.

I've also renamed the `CellOp` to `InstanceOp`, and added a (hopefully) cleaner interface for primitives to minimize code duplication.

### CODE REVIEW

File: lib/Dialect/Calyx/CalyxOps.cpp [hunk 1]
```diff
@@ -401,65 +412,78 @@ GroupDoneOp GroupOp::getDoneOp() {
...
 
 /// Lookup the component for the symbol. This returns null on
 /// invalid IR.
-ComponentOp CellOp::getReferencedComponent() {
+ComponentOp InstanceOp::getReferencedComponent() {
   auto program = (*this)->getParentOfType<ProgramOp>();
   if (!program)
     return nullptr;
 
   return program.lookupSymbol<ComponentOp>(componentName());
```
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:441]: I was wondering whether this is safe? The `ComponentOp` parent of the `InstanceOp` has trait `IsolatedFromAbove` which makes me think that accessing the enclosing ProgramOp's symbol table is unsafe. I am still new to MLIR so maybe I just misunderstand the legality implications of `IsolatedFromAbove`!

Comment (stephenneuendorffer) [reply lib/Dialect/Calyx/CalyxOps.cpp:441]: IsolatedFromAbove only refers to SSA values.  Indeed, one of the reasons behind the need for symbol tables is to provide an explicit mechanism for regions that are isolatedFromAbove (like a region inside a function definition) to refer to other code (like another function definition in a call).  see https://mlir.llvm.org/docs/Traits/#isolatedfromabove

**Review by mikeurbach — APPROVED** [review]
Makes sense to me.

**Comment by cgyurgyik** [comment]
> Makes sense to me.

I'm going to spend a bit more time looking into a cleaner way than just throwing in a function. It looks like an `OpInterface` may be better, as suggested [here](https://llvm.discourse.group/t/changing-the-default-argument-for-child-classes/4098/3). If this is blocking the Memory primitive, let me know and I'll merge this and work on a new PR.

**Comment by mikeurbach** [comment]
Not blocking me, unless someone else is looking for a Memory primitive.

**Comment by cgyurgyik** [comment]
I won't have time to add an interface in the short term, so I've created #1597 for now.

**Review by mikeurbach — APPROVED** [review]
The trait LGTM. I will try using it for a memory primitive, and if it makes sense I might try to push it to be an interface.

**Pull request merged**